crates/bevy_math/src/catmull_rom.rs

```diff
@@ -0,0 +1,149 @@
+use glam::{Vec2, Vec3, Vec3A};
+
+use std::{
+    fmt::Debug,
+    iter::Sum,
+    ops::{Add, Mul, Sub},
+};
+
+/// A point in space of any dimension that supports addition and multiplication.
+pub trait Point:
+    Mul<f32, Output = Self>
+    + Add<Self, Output = Self>
+    + Sub<Self, Output = Self>
+    + Add<f32, Output = Self>
+    + Sum
+    + Default
+    + Debug
+    + Clone
+    + PartialEq
+    + Copy
+{
+}
+impl Point for Vec3 {}
+impl Point for Vec3A {}
+impl Point for Vec2 {}
+
+/// A spline that passes through all of its points.
+///
+/// The Catmull-Rom spline can be controlled with its `points`, and the `tension`. The `tension`
+/// determines how closely the spline follows the linear path between points.
+#[derive(Clone, Debug, Default, PartialEq)]
+pub struct CatmullRom<P: Point> {
+    points: Vec<P>,
+    tension: f32,
+    segments: Vec<Segment<P>>,
+}
+
+/// Represents a segment of a Catmull-Rom spline, used to hold precomputed coefficients for fast
+/// interpolation. A segment is composed of a path and its four nearest points which influence it.
+#[derive(Clone, Debug, Default, PartialEq)]
+pub struct Segment<P: Point> {
+    coeff: [P; 4],
+}
+
+impl<P: Point> CatmullRom<P> {
+    /// Construct a new Catmull-Rom spline
+    pub fn new(points: impl Into<Vec<P>>, tension: f32) -> Result<Self, CatmullRomError> {
+        let points = points.into();
+        let segments = Self::compute_segment_coefficients(&points, tension)?;
+        Ok(Self {
+            points,
+            tension,
+            segments,
+        })
+    }
+
+    /// Compute the position coordinate at `t` along the spline.
+    ///
+    /// Note that `t` varies from `0..=(n_points - 3)`.
+    pub fn position(&self, t: f32) -> P {
+        let (segment, t) = self.segment(t);
+        let [a, b, c, d] = segment.coeff;
+        a + b * t + c * t.powi(2) + d * t.powi(3)
+    }
+
+    /// Compute the instantaneous velocity vector at `t` along the spline.
+    ///
+    /// Note that `t` varies from `0..=(n_points - 3)`.
+    pub fn velocity(&self, t: f32) -> P {
+        let (segment, t) = self.segment(t);
+        let [_, b, c, d] = segment.coeff;
+        b + c * 2.0 * t + d * 3.0 * t.powi(2)
+    }
+
+    /// Compute the instantaneous acceleration vector at `t` along the spline.
+    ///
+    /// Note that `t` varies from `0..=(n_points - 3)`.
+    pub fn acceleration(&self, t: f32) -> P {
+        let (segment, t) = self.segment(t);
+        let [_, _, c, d] = segment.coeff;
+        c * 2.0 + d * 6.0 * t
+    }
+
+    /// Returns the [`Segment`] and local `t` value given a spline's global `t` value.
+    fn segment(&self, t: f32) -> (&Segment<P>, f32) {
```

Comment:
Is this intended to be private?

Reply:
Yeah, it seemed like more of an implementation detail than anything. Segments are just a collection of coefficients used for fast interpolation.